executor/common_linux.h

```diff
@@ -1129,25 +1137,26 @@ static long syz_mount_image(volatile long fsarg, volatile long dir, volatile uns
 			debug("syz_mount_image: pwrite[%u] failed: %d\n", (int)i, errno);
 		}
 	}
-	snprintf(loopname, sizeof(loopname), "/dev/loop%llu", procid);
-	loopfd = open(loopname, O_RDWR);
-	if (loopfd == -1) {
-		err = errno;
-		goto error_close_memfd;
-	}
-	if (ioctl(loopfd, LOOP_SET_FD, memfd)) {
-		if (errno != EBUSY) {
-			err = errno;
-			goto error_close_loop;
-		}
-		ioctl(loopfd, LOOP_CLR_FD, 0);
-		usleep(1000);
-		if (ioctl(loopfd, LOOP_SET_FD, memfd)) {
-			err = errno;
-			goto error_close_loop;
+
+	srand(procid + syscall(SYS_gettid));
+	loopdevno = (unsigned int)rand();
```

Comment:
What's wrong with the current code?
This new code looks non-deterministic, slower and leaks loop devices (? we don't do LOOP_CLR_FD on it).

Reply:
When looking at the control flow with the test, I noticed that the EBUSY that is returned was from this chunk
of code. LOOP_SET_FD fails, LOOP_CLR_FD is invoked, and LOOP_SET_FD fails again with EBUSY. Some possible for reasons seemed to be.

- there is a mount associated with this loop device already
- there is a different thread in the same process that did a LOOP_SET_FD after the LOOP_CLR_FD

In both cases, it seemed like it might be ideal to just find a "free" loopdevice and use it rather than contest over an
existing one. The change tries to do that. The idea was to do something as follows :-

1. Try to add a loopdevice
2. Try to attach the file to loopdevice. If successful, continue to use the loopdevice
3. If that fails, don't try to contest, just retry [1] with a different loodevice number

If LOOP_SET_FD fails, is there a need to LOOP_CLR_FD? I assumed that it was done to try and free the loopdevice
(and not related to leaking). Is my understanding incorrect/incomplete?

Yes, if this makes it hard to repro crashes then a different implementation is needed. As this is the first point where the testcase fails, maybe this should be fixed first.

Reply:
> LOOP_SET_FD fails, LOOP_CLR_FD is invoked, and LOOP_SET_FD fails again with EBUSY. Some possible for reasons seemed to be.

Why does the first LOOP_SET_FD fail? What is the error code? It seems that's where the problems start.
Does `sudo losetup /dev/loop0 somefile` work from shell? Does it work for other /dev/loopN?

Trying all loop devices will introduce very bad inference between test processes. Such inference causes non-reproducible crashes and non-reproducible coverage.
Also it can mask some bugs that we have. E.g. we fail to free up the loop devices. During limited testing it will work, because we will just use loop0, loop1, loop2 and so on and we will have enough of them to pass tests. But during actual fuzzing, we will run out of available loop devices very quickly and then actually no meaningful testing will happen. And this all will be completely masked and nobody will have a slightest worry that something is not working. Moreover passing tests will actually make everybody think that everything is working fine.
We need to understand what is the problem there.

Reply:
Hmm, I see. I did a bit more digging into this. I modified my local test to have the following system call sequence.

```
mkdir(...)
syz_mount_image(...)
umount2(...)
```
/thr was disabled(temporarily) as:
```
for _, threaded := range []bool{false, false} {
```

The results were
```
The results are as follows :-
syz_mount_image_ext4  C               : OK
syz_mount_image_ext4 /repeat C        : OK
syz_mount_image_ext4  C               : OK
syz_mount_image_ext4 /repeat C        : OK
syz_mount_image_ext4 none             : FAIL: run 0: call 1 is not executed
syz_mount_image_ext4 none/cover       : OK
syz_mount_image_ext4 none C           : OK
syz_mount_image_ext4 none/repeat      : FAIL: run 0: call 1 is not finished
syz_mount_image_ext4 none/repeat/cover: OK
syz_mount_image_ext4 none/repeat C    : OK
syz_mount_image_ext4 none             : FAIL: run 0: call 1 is not finished
syz_mount_image_ext4 none/cover       : OK
syz_mount_image_ext4 none C           : OK
syz_mount_image_ext4 none/repeat      : FAIL: run 0: call 1 is not executed
syz_mount_image_ext4 none/repeat/cover: OK
syz_mount_image_ext4 none/repeat C    : OK
ok: 12, broken: 0, skip: 2, fail: 4
tests failed
```

Before looking into the failures related to sandbox, what is the ideal behavior of this test under /thr? If the test has umount2, /thr tests will fail when umount2 is executed before syz_mount_image. If the test does not have umount2, then the loop device will not be free(i.e. even the /repeat will fail with EBUSY)

Reply:
In short, it seems like if a loop device is associated with a mountpoint, LOOP_CLR_FD does not free the loopdevice and all calls to LOOP_SET_FD will return EBUSY. This test program demonstrates that(https://paste.ubuntu.com/p/FKfdXdPdcC/)

Output
```
root@syzkaller:~# ./test
[   47.446144] EXT4-fs (loop0): mounted filesystem with ordered data mode. Opts: ,errors=continue
memfd=3
first LOOP_SET_FD failed: Device or resource busy
LOOP_CLR_FD returned 0
second LOOP_SET_FD failed: Device or resource busy
```

Reply:
These are probably due to timeouts:
`
syz_mount_image_ext4 none             : FAIL: run 0: call 1 is not executed
syz_mount_image_ext4 none/repeat      : FAIL: run 0: call 1 is not finished
syz_mount_image_ext4 none             : FAIL: run 0: call 1 is not finished
syz_mount_image_ext4 none/repeat      : FAIL: run 0: call 1 is not executed
`

> Before looking into the failures related to sandbox, what is the ideal behavior of this test under /thr? If the test has umount2, /thr tests will fail when umount2 is executed before syz_mount_image.

The intention of threaded is not to reorder syscalls. That's collide mode. The threaded mode allows to deal with blocking syscalls. Namely, syscalls are started in order, and we wait for their completion and when the current syscall ends, we start the next one. If a syscall blocks (does not return for too long), we start then next one without waiting for the current one.
So the expected behavior with threaded is the same as without threaded mode.
But the mount syscall can take too long and be falsely considered as "blocked". That may interfere with test execution. But threaded mode is what is used during fuzzing, so if it fails in tests, it also fails during fuzzing.
There is no good generic solution for this (strictly saying unsolvable problem as we don't know if it has blocked or just slow). @xairy observed similar problem in USB testing:
https://github.com/google/syzkaller/pull/982/files#diff-f855ddaf59997a8644b3c1c87dd267f1
We can do something similar here too.

Reply:
> In short, it seems like if a loop device is associated with a mountpoint, LOOP_CLR_FD does not free the loopdevice and all calls to LOOP_SET_FD will return EBUSY. This test program demonstrates that(https://paste.ubuntu.com/p/FKfdXdPdcC/)
> 
> Output
> 
> ```
> root@syzkaller:~# ./test
> [   47.446144] EXT4-fs (loop0): mounted filesystem with ordered data mode. Opts: ,errors=continue
> memfd=3
> first LOOP_SET_FD failed: Device or resource busy
> LOOP_CLR_FD returned 0
> second LOOP_SET_FD failed: Device or resource busy
> ```



Yikes!
I guess we hit this path:

```
static int loop_clr_fd(struct loop_device *lo)
{
...
	/*
	 * If we've explicitly asked to tear down the loop device,
	 * and it has an elevated reference count, set it for auto-teardown when
	 * the last reference goes away. This stops $!~#$@ udev from
	 * preventing teardown because it decided that it needs to run blkid on
	 * the loopback device whenever they appear. xfstests is notorious for
	 * failing tests because blkid via udev races with a losetup
	 * <dev>/do something like mkfs/losetup -d <dev> causing the losetup -d
	 * command to fail with EBUSY.
	 */
	if (atomic_read(&lo->lo_refcnt) > 1) {
		lo->lo_flags |= LO_FLAGS_AUTOCLEAR;
		mutex_unlock(&loop_ctl_mutex);
		return 0;
	}
```

Reply:
Humm... we should have been unmounted everything in remove_dir function. Do we do it in a wrong order? Or did it fail for some reason?